### DISCUSSION TITLE
use Traefik as the interceptor

### DISCUSSION THREAD

**Discussion by trondhindenes** [kedacore/http-add-on#213] [Ideas]
Traefik supports plugins (https://doc.traefik.io/traefik/plugins/) and it seems to me that for shops already using Traefik as the ingress controller, this is the perfect place to put the interceptor, instead of having to pass http traffic thru yet another layer.

It would be awesome if an option for deploying the http-scaler interceptor was to simply hook in a traefik plugin.

NB: I have not investigated the traefik plugin api in detail, so I don't know if this is technically feasable or not.

**Comment by arschles** [comment]
@trondhindenes is Traefik Pilot required to use a plugin? There seems to be a connection between the two things, but I'm not fully understanding the extent of it. Regardless, I'm not familiar with the metrics that Traefik exposes in its prometheus interface, but do you think we can plug KEDA into those? If so, the only other matter would be figuring out how to hold the request when the app is scaled to zero. I have some ideas on that as well.

**Reply by trondhindenes** [reply]
I'm not 100% on top of the newer Traefik stuff myself - was an active user in a previous job (and hope to implement it again in my current gig). Regardless, I think Pilot is the "app store", but actually installing plugins can be done using regular startup parameters to the binary (see https://doc.traefik.io/traefik-pilot/plugins/overview/). But I could be wrong, Traefik Pilot might be a "hard" requirement for the plugin stuff to work.

I was thinking that a "keda plugin" could do whatever you needed, such as expose metrics in the format that the http-scaler accepts, and possibly also holding the request while waiting for scale-up to occur. But again, I haven't looked into the Traefik api to find out what's possible.

I just thought i'd be cool to have http-scaler integrate with existing ingress-controllers instead of adding another component, especially since it would sit bang in the middle of a very critical path (http request).

As far as prometheus metrics go: Traefik exposes metrics on a "ingress by ingress" level if memory serves. Maybe that's all you need - no need for a custom plugin? I'm sure the Traefik maintainers would love to discuss this scenario too, they're awesome and very friendly people.

**Reply by arschles** [reply]
@trondhindenes thanks for that information! I do think that most of what we need could come from the prometheus (or other stats) interface. The only remaining part I see is that we need to toggle Traefik to forward to the interceptor when there are 0 replicas, and to forward directly to the app when there are >0. That can be done with a control loop.

>I just thought i'd be cool to have http-scaler integrate with existing ingress-controllers instead of adding another component, especially since it would sit bang in the middle of a very critical path (http request).

I, too, think this would be cool 😄

**Reply by tomkerkhove** [reply]
While I get the sentiment, the tricky thing is - We can do it for Traefik but then we scope our userbase to just that.

If we add support for this, I'm hoping it would be a more convenient way to use it and not fully scope it to just that. If that's the case, then it's fine for me!

**Reply by trondhindenes** [reply]
yup - I get that. Probably not a smart thin to take a hard requirement on a single flavor of ingress controller - not even the most popular one.

**Reply by tomkerkhove** [reply]
And another thing is, we don't want to enforce using ingresses on people, because maybe they want to use it for service-to-service scaling.

**Reply by arschles** [reply]
> We can do it for Traefik but then we scope our userbase to just that.

@tomkerkhove agreed, that's not a great outcome. my thought is that if we hook into traefik's prom. metrics and modify an `Ingress` to handle scale-from-zero cases, then we leave ourselves open for other implementations. WDYT?

**Comment by linkvt** [comment]
Closing as outdated — this predates several releases and either got resolved since or never reached a conclusion. See current docs (https://keda.sh/http-add-on/0.15/) or feel free to open a new discussion if still relevant.
